README
Date-Holidays-EnglandWales version 0.01
=======================================

This module can be used to determine whether a given date is a bank 
holiday in England and Wales. 

It is not suitable for determining bank holidays in Scotland or 
Northern Ireland.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Time::Piece
  DateTime::Event::Easter

COPYRIGHT AND LICENCE

Copyright (C) 2010 by Jason Clifford

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. This means under either the 
Artistic License or the GNU GPL version 2 or later.




Changes
Revision history for Perl extension Date::Holidays::EnglandWales.

0.09  Mon Sep 12 10:49:00 2022
    - Bank holiday for the State Funeral of Queen Elizabeth II

0.08  Sat Apr 16 13:11:00 2022
    - Move Spring Bank Holiday and add Platinum Jubilee

0.07  Sat Apr 16 17:50:00 2011
    - Drop Time::Piece and use DateTime instead - courtesy of David
        Precious (http://www.preshweb.co.uk/)

0.06  Tue Nov 23 15:22:00 2010
    - Add 29 April 2011 Bank Holiday for Wedding of Prince William 
        and Kate Middleton

0.05  Mon Feb  1 14:45:00 2010
    - If the 27th Dec is either Mon or Tue it means Christmas Day was 
      Saturday or Sunday so both Mon and Tue are bank holidays
        Thanks again to Lee Johnson for his eagle eye :)
      This also means the Chrismas Day bank holiday can be on 26th.

0.04  Mon Feb  1 12:20:00 2010
    - Boxing Day holiday can be either Mon or Tue if it's the 28th.

0.03  Mon Feb  1 12:10:00 2010
    - Boxing Day holiday moves to Tue is Christmas Day is Saturday or Sunday.
        Thanks for Lee Johnson for catching this one
    - Corrected dependency to include Time::Seconds

0.02  Fri Jan 29 09:59:00 2010
    - Doesn't require perl 5.10 - set it to 5.8 which should be enough.
    - Correct bug in Easter functions
    - Remove print statement

0.01  Thu Jan 28 17:48:21 2010
	- initial release



lib/Date/Holidays/EnglandWales.pm
package Date::Holidays::EnglandWales;

use 5.008000;
use strict;
use warnings;

use DateTime;
use DateTime::Event::Easter;

require Exporter;
our @ISA = qw(Exporter);
our @EXPORT = qw( is_holiday is_uk_holiday );
our $VERSION = '0.09';


=head1 NAME

Date::Holidays::EnglandWales - Determines bank holidays

=head1 SYNOPSIS

  use Date::Holidays::EnglandWales;
  
  my ($year, $month, $day) = (localtime)[ 5, 4, 3 ];
  $year  += 1900;
  $month += 1;

  my $isodate = "$year-$month-$day";

  print "It's a bank holiday" if is_holiday($year, $month, $day);
  print "Sleep in late!" if Date::Holidays::EnglandWales->is_holiday($isodate);

=head1 DESCRIPTION

Date::Holidays::EnglandWales returns true is a given date is a bank 
holiday in England and wales.

The date can be passed as year, month, day or as an ISO formatted date.

This module uses a simple set of rules to determine whether a date is a 
holiday rather than a static list. This means it isn't limited to just
the next few years and wont require maintainance unless the rules are
changed.

It knows about the proposed bank holiday for the Queen's Diamond Jubilee.

=head2 EXPORT

is_holiday
is_uk_holiday

=head1 DEPENDANCIES 

This module uses the following modules which you can get from CPAN.

Time::Piece
DateTime::Event::Easter

=head1 SEE ALSO

Bank Holidays in England and Wales are determined by the Banking and 
Financial Dealings Act 1971 and by Royal Proclamation. This means that it
is possible that they may change from year to year although in practice
this does not happen.

If you need to be absolutely sure about a date check with the DTI whose
website is http://www.dti.gov.uk/

=head1 AUTHOR

Jason Clifford, E<lt>jason@ukfsn.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 by Jason Clifford

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. That means either the Artistic
License or the GNU GPL version 2 or later.

=cut

sub is_holiday { goto &is_uk_holiday; }

sub is_uk_holiday {
    my ($year, $month, $day) = @_;
    ($year, $month, $day) = split m{[-/]}, $year unless $month;
    
    my $dt = DateTime->new(year => $year, month => $month, day => $day)
        or die "Failed to create DateTime object for $year-$month-$day"
            . " - invalid date?";

    if ( $month == 1 ) {
        return "New Years Day" if (
            ( $day == 1 && $dt->day_of_week <= 5 ) || 
            ( $day == 2 && $dt->day_of_week == 1 ) ||
            ( $day == 3 && $dt->day_of_week == 1 ) );
        return undef;
    }

    if ( $year == 2011 && $month == 4 ) {
        return "Wedding of Prince William and Kate Middleton" if $day == 29;
    }

    if ( $month == 5 ) {
        return "Early May Bank Holiday" if (
            ( $day >= 1 && $day <= 7 && $dt->day_of_week == 1 ) );

        return "Spring Bank Holiday" if (
            $year != 2012 && $year != 2022 &&
            ( $day >= 25 && $dt->day_of_week == 1 ) );
    }

    if ( $year == 2012 && $month == 6 ) {
        return "Spring Bank Holiday" if $day == 4;
        return "Queen's Diamond Jubilee" if $day == 5;  # Subject to Her Maj not dying first but I cannot be bothered to code for that.
        return undef;
    }

    if ( $year == 2022 && $month == 6) {
        return "Spring Bank Holiday" if $day == 2;
        return "Queen's Platinum Jubilee" if $day == 3; # She's still here
    }

    if ( $year == 2022 && $month == 9 ) {
	return "State Funeral of Queen Elizabeth II" if $day == 19; # May she rest in peace.
    }

    if ( $month == 8 ) {
        return "Summer Bank Holiday" if (
            $day >= 25 && $dt->day_of_week == 1 );
        return undef;
    }

    if ( $month == 12 ) {
        return "Christmas Day Bank Holiday" if (
            ( $day == 25 && $dt->day_of_week <= 5 ) ||
            ( $day == 26 && $dt->day_of_week == 1 ) ||
            ( $day == 27 && $dt->day_of_week == 1 ) );

        return "Boxing Day" if (
            ( $day == 26 && $dt->day_of_week <= 5 ) ||
            ( $day == 27 && $dt->day_of_week <= 2 ) ||
            ( $day == 28 && $dt->day_of_week <= 2 ) );

        return undef;
    }

    return "Good Friday" if &_GoodFriday($year, $month, $day);

    # We need to check if the day before the day in question is Easter Sunday
    my $e = $dt->clone->subtract(days => 1);
    return "Easter Monday" if &_Easter($e->year, $e->month, $e->day);

}

sub _Easter {
    my ($year, $month, $day) = @_;

    my $dt = DateTime->new( year => $year, month => $month, day => $day );
    my $easter_sunday = DateTime::Event::Easter->new();

    return $easter_sunday->is($dt);
}

sub _GoodFriday {
    my ($year, $month, $day) = @_;

    my $dt = DateTime->new( year => $year, month => $month, day => $day );
    my $good_friday = DateTime::Event::Easter->new(day=>'Good Friday');

    return $good_friday->is($dt);
}

1;
__END__
